generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt

```diff
@@ -0,0 +1,441 @@
+                val envelope = Envelope(minPoint, maxPoint)
+                downloadArea.geometry = envelope
+            }
+        }
+
+        // set onclick listener for the takeMapOfflineButton
+        takeMapOfflineButton.setOnClickListener {
+            // if the downloadArea's geometry is not null
+            downloadArea.geometry?.let { geometry ->
+                // create an OfflineMapJob
+                val offlineMapJob = createOfflineMapJob(map, geometry)
+                // start the OfflineMapJob
+                startOfflineMapJob(offlineMapJob)
+                // show the progress dialog
+                progressDialog.show()
+                // disable the button
+                takeMapOfflineButton.isEnabled = false
+            }
+        }
+
+        // start observing the worker's progress and status
+        observeWorkStatus()
+    }
+
+    /**
+     * Creates and returns a new GenerateOfflineMapJob for the [map] and its [areaOfInterest]
+     */
+    private fun createOfflineMapJob(
+        map: ArcGISMap,
+        areaOfInterest: Geometry
+    ): GenerateOfflineMapJob {
+        // check and delete if the offline map package file already exists
+        File(offlineMapPath).deleteRecursively()
+        // specify the min scale and max scale as parameters
+        val maxScale = map.maxScale ?: 0.0
+        var minScale = map.minScale ?: 0.0
+        // minScale must always be larger than maxScale
+        if (minScale <= maxScale) {
+            minScale = maxScale + 1
+        }
+        // set the offline map parameters
+        val generateOfflineMapParameters = GenerateOfflineMapParameters(
+            areaOfInterest,
+            minScale,
+            maxScale
+        ).apply {
+            // set job to cancel on any errors
+            continueOnErrors = false
+        }
+        // create an offline map task with the map
+        val offlineMapTask = OfflineMapTask(map)
+        // create an offline map job with the download directory path and parameters and
+        // return the job
+        return offlineMapTask.createGenerateOfflineMapJob(
+            generateOfflineMapParameters,
+            offlineMapPath
+        )
+    }
+
+    /**
+     * Starts the [offlineMapJob] using OfflineJobWorker with WorkManager. The [offlineMapJob] is
+     * serialized into a json file and the uri is passed to the OfflineJobWorker, since WorkManager
+     * enforces a MAX_DATA_BYTES for the WorkRequest's data
+     */
+    private fun startOfflineMapJob(offlineMapJob: GenerateOfflineMapJob) {
+        // create a temporary file path to save the offlineMapJob json file
+        val offlineJobJsonPath = getExternalFilesDir(null)?.path +
+            getString(R.string.offlineJobJsonFile)
+
+        // create the json file
+        val offlineJobJsonFile = File(offlineJobJsonPath)
+        // serialize the offlineMapJob into the file
+        offlineJobJsonFile.writeText(offlineMapJob.toJson())
```

Comment:
Is there a need to write the serialized job to a file? You could just pass the JSON string directly to the work request.

Reply:
There is a [MAX_DATA_BYTES](https://developer.android.com/reference/kotlin/androidx/work/Data#MAX_DATA_BYTES()) size limit that is enforced on the payloads for `Data` that we add to a `WorkRequest`. It's about ~10Kb and often the JSON string exceeds that.

This is added to the function level comments too - https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/blob/5911bbe08a428ed06136864d5b054092165bc8c6/generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt#L229-L234